gateway/handlers/reader.go

```diff
@@ -71,3 +93,24 @@ func MakeFunctionReader(metricsOptions metrics.MetricOptions, c client.ServiceAP
 		w.Write(functionBytes)
 	}
 }
+
+func getReplicaInfo(c client.ServiceAPIClient, services []swarm.Service, ctx context.Context) map[string]int {
+	// Begin replica info section
+	taskFilter := filters.NewArgs()
+
+	taskFilter.Add("_up-to-date", "true")
+	tasks, err := c.TaskList(ctx, types.TaskListOptions{Filters: taskFilter})
+	if err != nil {
+		log.Println(err)
+	}
+
+	running := map[string]int{}
+	for _, task := range tasks {
+		if task.Status.State == swarm.TaskStateRunning {
```

Comment:
Wondering if node state should be taken into account; if a node is not reachable, task states are not updated (i.e., the task state then reflects the last _known_ state); see https://github.com/docker/cli/blob/master/cli/command/service/list.go#L118-L120

```go
if _, nodeActive := activeNodes[task.NodeID]; nodeActive && task.Status.State == swarm.TaskStateRunning {
	running[task.ServiceID]++
}
```

Be aware that in all cases, this state would likely be on a "best effort" base; managers (and raft store) have the _desired state_ in store, and the nodes report back what the actual state is, so there may be a latency.

Reply:
Thanks for looking at this. I thought task states would have been factored into whether a task was ready or not. I asked Eric to try without it and I think we sped things up by a large factor.

Sounds like they might need to come back in then @ericstoekl ?

Reply:
Yeah, sounds like nodes need to come back. Stefan Prodhan mentioned this as well during the contributors meeting two weeks back. It might be the only way to reliably get replica status.

Reply:
I can double check if you want now that SF starts to wake up 😅

Reply:
@alexellis There is an issue with having the function filter by `NodeList`, because the client object we have is of type `client.ServiceAPIClient`, which does note implement the `NodeList()` function:

https://github.com/docker/engine-api/blob/4290f40c056686fcaa5c9caf02eac1dde9315adf/client/interface.go#L102

Before [this commit](https://github.com/openfaas/faas/commit/0248a5ea1d9cb412e64df70a13b1337f8a2b625e) `c` was a `client.Client` object, which had the `NodeList()` method available.

Reply:
Hey Eric.. we can work around this by passing in the equivalent interface / struct through from the calling code. You just need the Node API client right? This one? https://github.com/docker/engine-api/blob/4290f40c056686fcaa5c9caf02eac1dde9315adf/client/interface.go#L94